staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go

```diff
@@ -483,11 +485,14 @@ func (cnc *CloudNodeController) getNodeModifiersFromCloudProvider(
 		}
 	}
 
-	// If user provided an IP address, ensure that IP address is found
-	// in the cloud provider before removing the taint on the node
-	if nodeIP, ok := ensureNodeProvidedIPExists(node, instanceMeta.NodeAddresses); ok && nodeIP == nil {
+	// If user provided a node IP address, ensure that IP address is found
+	// in the cloud provider and place it first in the node addresses list
+	// before removing the taint on the node
+	nodeAddresses := preferProvidedNodeIP(node, instanceMeta.NodeAddresses)
+	if nodeAddresses == nil {
```

Comment:
This is side-effecty and non-obvious, would be better to return an explicit bool or error when we don't find the address wouldn't it? Can `instance.NodeAddresses` ever be nil? If it is, there could be multiple ways that `nodeAdresses` becomes nil which makes this even more like guesswork

Reply:
NodeAddresses *should* never be nil, but now that you've said it I don't think it's good practise to assume that here and we should handle it.

I agree it's side-effecty, although returning the input parameter was my attempt to make this more obvious. The previous iteration just returned a bool, and I didn't think that was obvious at all that it had a side-effect. Perhaps if it was in the name: `modifyNodeAddressesForPreferredNodeIP() -> bool(ok)`? That would also have the advantage of gracefully handling nil NodeAddresses 🤔

Reply:
> returning the input parameter was my attempt to make this more obvious

This is good, and there's no reason not to return two return variables right? You could return the list and a bool, or a list and an error, I could see either of those making sense here depending how you look at it